src/guest_config/src/guest_config_compiler_bin.rs

```diff
@@ -0,0 +1,366 @@
+                output_file: String::from(GUESTCONFIG_DEFAULT_OUTPUT_FILENAME),
+            };
+
+            match compile(&args).unwrap_err() {
+                GuestConfigCompilerError::TemplateConfiguration(
+                    GuestConfigurationError::JsonError(_),
+                ) => {} // success
+                err => {
+                    panic!(
+                        "Expected GuestConfigCompilerError::GuestConfigurationError::JsonError \
+                         error. Receieved [{:?}]",
+                        err
+                    )
+                }
+            }
+        }
+
+        // test a successful compilation
+        {
+            // create usable cpuid object
+            let cpuid_config = Cpuid::kvm_get_supported_cpuid().expect("Failed to get CPUID");
+            let cpuid_file = Builder::new()
+                .prefix("cpuid-config")
+                .suffix(".json")
+                .tempfile()
+                .unwrap();
+            let in_file = Builder::new()
+                .prefix("cpu-json-template-config")
+                .suffix(".json")
+                .tempfile()
+                .unwrap();
+            let out_file = Builder::new()
+                .prefix("cpu-bin-template-config")
+                .suffix(".json")
+                .tempfile()
+                .unwrap();
+
+            // Write CPUID json file
+            let cpuid_path = cpuid_file.path().to_str().unwrap();
+            let cpuid_json = serde_json::to_string_pretty(&cpuid_config).unwrap();
+            fs::write(cpuid_path, cpuid_json).expect("Unable to write JSON test file");
+
+            let configuration_json = get_correct_json_input(Some(cpuid_path.to_string()));
+            if let Err(err) = fs::write(in_file.path(), &configuration_json) {
+                panic!("Error writing test json file \n{:?}", err);
+            }
+
+            let arguments = Arguments {
+                input_file: String::from(in_file.path().to_str().unwrap()),
+                output_file: String::from(out_file.path().to_str().unwrap()),
+            };
+
+            // do the compilation & check for errors
+            match compile(&arguments) {
+                Ok(_) => {}
+                Err(err) => {
+                    error!(
+                        "Failed to parse request, json structure as \n{}",
+                        configuration_json
+                    );
+                    panic!("Unexpected error running compiler \n{:?}", err)
+                }
+            }
+        }
+    }
+
+    // test helper for generating correct JSON input data
+    fn get_correct_json_input(arch_features_file_path: Option<String>) -> String {
+        let file_path: String;
+        match arch_features_file_path {
+            None => file_path = String::from("/tmp/cpuid-test.json"),
+            Some(arch_file_path) => file_path = arch_file_path,
+        }
```

Comment:
This can be simplified.
```suggestion
        let file_path = match arch_features_file_path {
            None => String::from("/tmp/cpuid-test.json"),
            Some(arch_file_path) => arch_file_path,
        };
```